### PR TITLE
[JAX] Rewrite the Format of FP8 Meta and Remove unused ShardingTypes.

### PR DESCRIPTION
**Pull request by mingxu1067**
# Description

Reformatted FP8 meta to one set per tensor, removed `fp8_max` and `scale_inv` from the set of FP8 meta, and deleted unused functions and types.

Fixes # (issue)
To avoid unnecessary `slice` of FP8 meta then unblock pipeliner to re-schedule the collectives.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
The existing FP8 checkpoints would not be compatiable.

## Changes

Please list the changes introduced in this PR:

1. Reformat FP8 meta to be one-set-per-tensor.
2. Remove `fp8_max` and `scale_inv` from FP8 meta set.
3. Remove unused functions in fp8.py, like `update_fp8_metas`.
4. Remove unused `ShardingType` and `MajorShardingType`.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by nouiz** [comment]
Can you tell what speed up you see with this PR?

File: transformer_engine/jax/praxis/module.py
Comment (nouiz) [inline transformer_engine/jax/praxis/module.py:22]: This doesn't seem related to what the description of this PR does.
Can you add to the PR description to explain this part of the change?

File: transformer_engine/jax/__init__.py
Comment (nouiz) [inline transformer_engine/jax/__init__.py:26]: Also tell in the description that this function is now removed and not needed as fully automatic?
This is part of the API change this PR is doing.

File: tests/jax/test_helper.py
Comment (nouiz) [inline tests/jax/test_helper.py:78]: Why this loop? Just to test many rng?

File: transformer_engine/jax/praxis/module.py
Comment (mingxu1067) [reply transformer_engine/jax/praxis/module.py:22]: Done, add to the `Changes`.

File: transformer_engine/jax/__init__.py
Comment (mingxu1067) [reply transformer_engine/jax/__init__.py:26]: This function has not used for a while, after we migrated to `custom_partitioning`. Now FP8 meta is updated automatically in the backward, therefore no needed this function. Also added to `Changes`.

File: tests/jax/test_helper.py
Comment (mingxu1067) [reply tests/jax/test_helper.py]: Yes, for test different rngs

File: transformer_engine/jax/fp8.py
Comment (zlsh80826) [inline transformer_engine/jax/fp8.py:109]: I'm wondering if `dict` a better structure for maintaining the FP8 meta, we can use `jax.tree` for FP8 meta operations.

File: transformer_engine/jax/layernorm.py
Comment (zlsh80826) [inline transformer_engine/jax/layernorm.py:202]: Doesn't `[0:1]` equal to `[0]`?

File: transformer_engine/jax/fp8.py
Comment (zlsh80826) [inline transformer_engine/jax/fp8.py:112]: Do we need to add `_list` suffix? I think the type annotation is enough to represent it's data structure.

File: transformer_engine/jax/fp8.py
Comment (mingxu1067) [reply transformer_engine/jax/fp8.py:98]: Done

File: transformer_engine/jax/fp8.py
Comment (mingxu1067) [reply transformer_engine/jax/fp8.py:109]: The `list` and `dict` seem to have the simialr advantages here, since we eventially have to slice those meta to `amax` and `scale`. The difference here when to do so. If the structure is `list` then we split meta before calling VJP functions. If it is `dict`, then we could delay the split into VJP functions.

For `jax.tree`, since there are only one function could be shared between `amax` and `scale`, that is `converting to/from fm32`, then there are not much changes to take to advantages of `jax.tree`.

I think it might be fine to make it as `list` to be more straightforward at this moment, but we could discuss more offline.

File: transformer_engine/jax/layernorm.py
Comment (mingxu1067) [reply transformer_engine/jax/layernorm.py:202]: This is shape issue, if we do `[0]`, then the tensor shape would be `[]`, which is not compatiable to `cast` related custom calls.

File: transformer_engine/jax/fp8.py
Comment (mingxu1067) [reply transformer_engine/jax/fp8.py:112]: Adding `_list` here is to distinguish the `amax` tensors and `amax` containers, then we could recognize what the var contains by its name. Like `amax_list` means a list of multiple amax tensors and `amax` means an amax tensor.
But I have no strong thought here, we could discuss offline to make a final call.

**Comment by nouiz** [comment]
@denera As this PR is removed some old API not used, we need to have that documented in the next release. Is there a place we need to add them to be sure to be included in the next releases?

File: transformer_engine/jax/fp8.py
Comment (zlsh80826) [reply transformer_engine/jax/fp8.py]: It seems there is no special reason that the `dict` outperform `list` besides my personal preference. Let's just keep it as list. By the way, `jax.tree.map` also works on `list`, we can use that next time.

**Review by zlsh80826 — APPROVED** [review]
LGTM excepts I think we can add `jax.jit` for a few meta helper functions like `update_amax_list`, `update_fp8_scale` or `fm32_to_fp32_fun`.

**Comment by phu0ngng** [comment]
LGTM  👍. Would be interesting to see the diff in performance if any.

**Comment by mingxu1067** [comment]
Change to Draft for waiting internal verifiy

**Review by denera — APPROVED** [review]
LGTM! I'll merge after the `tansformer_engine/jax/flax/module.py` conflict is resolved.

**Pull request merged**
